include/mapnik/geometry_is_valid.hpp

```diff
@@ -106,6 +110,140 @@ inline bool is_valid(T const& geom)
     return detail::geometry_is_valid() (geom);
 }
 
+typedef boost::geometry::model::d2::point_xy<double> Point;
+typedef boost::geometry::model::segment<Point> Segment;
+
+template <typename T>
+std::vector<Segment> make_segment_vector(linear_ring<T> ring)
+{
+    std::vector<Segment> result;
+
+    if (ring.size() > 1) {
+        for (int i = 0; i < ring.size()-1; ++i)
+        {
+            Point a( ring.at(i).x, ring.at(i).y );
+            Point b( ring.at(i+1).x, ring.at(i+1).y );
+            Segment s(a, b);
+            result.push_back(std::move(s));
+        }
+    }
+
+    return result;
+
+}
+
+template <typename T>
+inline bool is_valid_rings(polygon<T> const& poly)
+{
+    // If there are no interior rings, then the polygon must be valid
+    if (poly.interior_rings.size() == 0) return true;
+
+    // If there is no exterior rings, and any interior rings have points
+    // then this is not a valid polygon
+    if (poly.exterior_ring.size() == 0) 
+    {
+        for (linear_ring<T> interior_ring : poly.interior_rings) 
```

Comment:
`for (linear_ring<T> interior_ring : poly.interior_rings) ` will lead to a copy of each ring. Best to ask for a `const&` to avoid unintentional copy since you don't need to modify the `interior_ring`. I recommend:

```
for (auto const& interior_ring : poly.interior_rings) 
````